### PR TITLE
Python script to replace ruby xmlschema generator

### PR DESCRIPTION
**Pull request by mjcarroll**
In conjunction with #884, this should allow us to remove the Ruby dependency from sdformat entirely.

I checked this against the output of the ruby generator, and all files matched with the exception of a small difference in physics:

```
diff --git a/physics.xsd b/physics.xsd
index af02efb..2b4df93 100644
--- a/physics.xsd
+++ b/physics.xsd
@@ -144,7 +144,8 @@
         cor = e                            [rigid]
 
         Combining rule e = 0,               e1==e2==0
-                         = 2*e1*e2/(e1+e2), otherwise]]>
+                         = 2*e1*e2/(e1+e2), otherwise
+      ]]>
                   </xsd:documentation>
                 </xsd:annotation>
                 <xsd:complexType>
```

### CODE REVIEW

**Comment by mjcarroll** [comment]
> CI is failing with a permission denied message

the power of `chmod +x`

**Comment by scpeters** [comment]
shall we remove the ruby script it is replacing?

**Comment by mjcarroll** [comment]
> shall we remove the ruby script it is replacing?

Good with me.  I noticed that the other PR doing this targeted main, so I will retarget this one.

**Comment by azeey** [comment]
This brought other changes over from `sdf13` when you retargetted it. https://github.com/gazebosim/sdformat/pull/1236 should help.

**Comment by mjcarroll** [comment]
> This brought other changes over from sdf13 when you retargetted it.

Rebased, thanks.

Note that we can't drop ruby entirely, because it will interfere with the CLI tests.

**Comment by Bi0T1N** [comment]
Is there a reason why you didn't considered #643?

**Comment by mjcarroll** [comment]
> Is there a reason why you didn't considered https://github.com/gazebosim/sdformat/pull/643?

There are other questions that needed to be resolved there.  I think that #643 is still relevant in the long term, I'm just trying to dump ruby dependencies as fast as possible here.

File: tools/xmlschema.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+def print_include_ref(element: ElementTree.Element, sdf_root_dir: str) -> List[str]:
+    """
+    Print include tag reference information
+    """
+    lines = []
+    filename = get_attribute(element, "filename")
+    if filename is not None:
+        sdf_path = os.path.join(sdf_root_dir, filename)
+        if os.path.exists(sdf_path):
```
Comment (azeey) [inline tools/xmlschema.py:125]: The ruby code would have thrown an exception or something if the file wasn't 
found. We're silently ignoring it here.

File: tools/xmlschema.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    lines = []
+
+    elem_name = get_attribute(element, "name")
+    elem_type = get_attribute(element, "type")
+
+    elements = element.findall("element")
+    attributes = element.findall("attribute")
+    includes = element.findall("include")
+
+    lines.extend(print_documentation(element))
```
Comment (azeey) [inline tools/xmlschema.py:234]: Should this be `elements` (plural) to match the ruby code?

File: tools/xmlschema.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+            lines.append("    </xsd:choice>")
+
+        for attribute_element in attributes:
+            lines.extend(indent_lines(print_attribute(attribute_element), 4))
+
+        lines.append("  </xsd:complexType>")
+        lines.append("</xsd:element>")
+    else:
+        if elem_type and is_std_type(elem_type):
+            elem_type = xsd_type_string(elem_type)
```
Comment (azeey) [inline tools/xmlschema.py:272]: From https://github.com/gazebosim/sdformat/blob/bfeaf0738be02742434092b97a55a1f32b5f2e2f/tools/xmlschema.rb#L232-L243

```suggestion
            elem_type = f' type={xsd_type_string(elem_type)}'
```

File: tools/xmlschema.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    parser.add_argument("--sdf-dir")
+    parser.add_argument("--output-dir")
+    args = parser.parse_args()
+
+    input_file = os.path.abspath(args.input_file)
+    output_lines = process(input_file, args.sdf_dir)
+    fname = os.path.splitext(os.path.basename(args.input_file))[0]
+    os.makedirs(args.output_dir, exist_ok=True)
+
+    output_file = os.path.join(args.output_dir, f"{fname}.xsd")
```
Comment (azeey) [inline tools/xmlschema.py:308]: Just an observation: the ruby script handles multiple top-level `<element>` tags 
while this doesn't, but I don't see any `.sdf` description files with multiple 
`<element>` tags.

```
xmlstarlet sel -t -v "count(/element)" sdf/**/*.sdf

```
Also, I believe xml allows only one top-level element.

**Review by azeey — COMMENTED** [review]
Looks pretty good! Just a few minor comments.

File: tools/xmlschema.py (hunk 2 above)
Comment (mjcarroll) [reply tools/xmlschema.py:234]: Not sure that I understand. The child elements are iterated below?

Comment (azeey) [reply tools/xmlschema.py:238]: Sorry, I was confused by the `elements` in the `_elem.elements["description"]` of the ruby code. Disregard.

File: tools/xmlschema.py (hunk 4 above)
Comment (azeey) [reply tools/xmlschema.py:312]: Agreed

**Review by azeey — APPROVED** [review]

**Comment by mjcarroll** [comment]
@scpeters do you know what would be causing:

```
Warning: HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK is set: not checking for outdated
dependents or dependents with broken linkage!
Error: Process completed with exit code 1.
##[debug]Finishing: Install base dependencies
```

In the actions build here?

**Comment by mjcarroll** [comment]
Nevermind, this is the root error:

```
ln -s ../Cellar/python@3.11/3.11.4_1/bin/2to3 2to3
rm /usr/local/bin/2to3
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /usr/local
Could not symlink bin/2to3-3.11
Target /usr/local/bin/2to3-3.11
already exists. You may want to remove it:
  rm '/usr/local/bin/2to3-3.11'

To force the link and overwrite all conflicting files:
  brew link --overwrite python@3.11

To list all files that would be deleted:
  brew link --overwrite --dry-run python@3.11

Possible conflicting files are:
/usr/local/bin/2to3-3.11 -> /Library/Frameworks/Python.framework/Versions/3.11/bin/2to3-3.11
/usr/local/bin/idle3.11 -> /Library/Frameworks/Python.framework/Versions/3.11/bin/idle3.11
/usr/local/bin/pydoc3.11 -> /Library/Frameworks/Python.framework/Versions/3.11/bin/pydoc3.11
/usr/local/bin/python3.11 -> /Library/Frameworks/Python.framework/Versions/3.11/bin/python3.11
/usr/local/bin/python3.11-config -> /Library/Frameworks/Python.framework/Versions/3.11/bin/python3.11-config
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FromPathLoader): loading /usr/local/opt/python@3.11/.brew/python@3.11.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/python@3.11.rb
```

**Comment by scpeters** [comment]
> @scpeters do you know what would be causing:
> 
> ```
> Warning: HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK is set: not checking for outdated
> dependents or dependents with broken linkage!
> Error: Process completed with exit code 1.
> ##[debug]Finishing: Install base dependencies
> ```
> 
> In the actions build here?

there was a fix in https://github.com/gazebosim/sdformat/pull/1271 waiting to be forward ported to `main` in https://github.com/gazebosim/sdformat/pull/1297

**Comment by mjcarroll** [comment]
> there was a fix in https://github.com/gazebosim/sdformat/pull/1271

I just cherry-picked the fix here.

**Pull request merged**